### PR TITLE
Allow use of project and file dir in rcfile, cwd, pythonpath.

### PR DESCRIPTION
**Pull request by rainyday**
Adds %p support to additional settings as well as %f for the directory of the current file. Allows pylint working directory to be customized for projects with nested packages.

[Review on Reviewable](https://reviewable.io/reviews/atomlinter/linter-pylint/91)

### CODE REVIEW

**Comment by Arcanemagus** [comment]
Mind rebasing this so it's just the commit making the changes?

**Comment by Arcanemagus** [comment]
Since you committed to your master branch instead of creating a branch for this feature that can be a tricky prospect, you will want to follow steps similar to the following:
```ShellSession
git checkout -b dir-config
git cherry-pick bd9dd45
git checkout master
git rebase -i HEAD~2
git push -f
git fetch --all
git merge remotes/AtomLinter/master
git push
git checkout dir-config
git rebase master
```
This will:
* Create a new branch
* Pull your commit doing the changes into it
* Checkout the master branch
* Remove the commits there that are conflicting with the history here. During the rebase you will want to comment out / remove _both_ commits so your master branch will be in the original state before any changes.
* Force pushing the fixed master back up to GitHub
* Pull down all the latest versions of the branches
* Merge the changes from the AtomLinter repository into your master so it is the latest version
* Push the updated master to GitHub
* Switch to the branch you started out creating
* Rebase it on the current master

_Hopefully_ I haven't screwed anything up in those instructions :stuck_out_tongue:

**Comment by rainyday** [comment]
Sorry about that, forgot I'd already forked awhile ago. My I'm using pylint 1.4.4 and it's not giving errors on empty files when run on cli either so I'm not sure if the test failure is a problem with my changes or a problem with the version of pylint.

**Comment by Arcanemagus** [comment]
Actually the test failures are due to a fix for a bug in one of the libraries that was causing it to report the wrong range. Would you like to just bump up the limit on that failing test or want me to push up a fix for that and then you would need to rebase? Either way is fine by me.

File: lib/main.coffee [hunk 1]
```diff
@@ -77,9 +88,10 @@ module.exports =
             '--output-format=text'
           ]
           if @rcFile
-            args.push "--rcfile=#{@rcFile}"
+            rcFile = @rcFile.replace(/%p/g, projDir).replace(/%f/g, path.dirname(file))
```
Comment (Arcanemagus) [inline lib/main.coffee]: Why the change in style to the call to `replace()`?

Comment (rainyday) [inline lib/main.coffee]: I figured that since we're just searching for a specific string to replace this probably performs better and requires one less character, not that it makes much difference.

Comment (Arcanemagus) [inline lib/main.coffee]: Earlier you used a substring, here you are using a regex, as far as i can tell a substring should work just fine here as well.

Comment (SpainTrain) [inline lib/main.coffee]: IMO, global replacement makes sense for `pythonPath` because we may want to add multiple directories in the project to `PYTHONPATH` e.g., `%p/foo:%p/bar`.  I cannot think of a use case for specifying project or file dir multiple times in the `rcFile` path.

File: lib/main.coffee [hunk 2]
```diff
@@ -62,9 +71,11 @@ module.exports =
...
         file = activeEditor.getPath()
         return helpers.tempFile path.basename(file), activeEditor.getText(), (tmpFilename) =>
-          projDir = @getProjDir(file)
-          cwd = projDir
-          pythonPath = @pythonPath.replace(/%p/g, projDir)
+          # default project dir to file directory if path cannot be determined
+          projDir = @getProjDir(file) or path.dirname(file)
+          cwd = @cwd.replace('%f', path.dirname(file)).replace('%p', projDir)
+          executable = @executable.replace('%p', projDir)
+          pythonPath = @pythonPath.replace('%p', projDir)
```
Comment (SpainTrain) [inline lib/main.coffee]: Please switch this back to using global replacement for `%p`, since we may want to add multiple subdirectories of the project dir to `PYTHONPATH`

```js
pythonPath = @pythonPath.replace(/%p/g, projDir)
```

Comment (SpainTrain) [inline lib/main.coffee]: In fact, I do this exact thing in my linter-pylint config =-D

File: lib/main.coffee (hunk 1 above)
Comment (rainyday) [inline lib/main.coffee]: I was using regex earlier as that's what was already being used but then changed it in a different commit. In light of @SpainTrain's comment I switched everything back to regex for the sake of consistency and flexibility.

**Comment by Arcanemagus** [comment]
Tests are passing and it looks like all comments have been resolved, I'm merging this in. You can yell at me later if you weren't done @SpainTrain :stuck_out_tongue:.
